### ISSUE TITLE
Should a purge event always be preceded by a delete event?

### ISSUE THREAD

**Issue by cmgrote** [odpi/egeria#548]
For event mapper implementations, should we always ensure that a `delete[Entity|Relationship]` event is sent prior to sending a `purge[Entity|Relationship]` event?

For repositories without soft deletes, I have been assuming we propagate the `purge` immediately, but perhaps this would create unexpected circumstances from a consumption standpoint?

**Comment by mandy-chessell** [comment]
This is not necessary.  Delete (and resume) are optional functions and so some repositories don't support them.

**Comment by cmgrote** [comment]
Perhaps I've misunderstood the response; from Graham:

> The design of metadata collections is that 'if a soft delete is supported, it must be performed prior to a purge'. Both the Atlas connector and the inMemory connector's implementations of purge both validate that the instance status is DELETED and if not they throw an exception. If a purge is issued without a prior delete (soft) then they will throw an EntityNotDeletedException or RelationshipNotDeletedException exception.

My assumption from above was that for a system that itself doesn't support soft delete, when a deletion occurs, that system does *not* need to first send a `delete[Entity|Relationship]` event -- but it sounds like systems that would receive such a notification will throw an error if they themselves *do* support soft delete but do not first receive such a `delete[Entity|Relationship]` (?)

**Comment by mandy-chessell** [comment]
Systems that do not support soft delete should not have to send the delete event before the purge event.   When they are sending these events, they are the master of the entity and control its lifecycle.  Systems receiving these events should just do as they are told and purge and/or delete as requested.

**Comment by cmgrote** [comment]
Thinking about it, sounds like the logic should be as follows:

- Check if my repository is the home repository for the instance in the event
   - Yes?  If I support delete and purge, and only receive a purge, then throw an exception.
   - No?  Just do as I'm told: if I receive a purge without a delete, do whatever I need to in order to purge.

(I'm assuming that since the home repository is the only place the instance can be modified, it's also the only place from which a delete / purge event can originate?  And therefore the home repository controls itself in terms of what it supports with respect to deletes vs purges.)

**Comment by grahamwallis** [comment]
I think it would be helpful if we are explicit about whether the scenarios are local or remote and the methods on the MetadataCollection interface that we are referring to. Let me try - let me know if this is goofy.

Local Operations
For local operations issued through the repository-services interface the metadataCollection implementation should reflect what the local repository connector supports; if the local repository does not support soft delete then the metadataCollection's purge methods will implement a purge that has not been preceded by a soft delete; whereas if the local repository does support soft delete the metadataCollection implementation of purgeEntity and purgeRelationship insist on a soft delete prior to the purge. Both the in-memory and atlas metadataCollections support soft delete so they are coded to enforce this; for these metadataCollections, the purge methods currently police this by checking whether (status == DELETED) and throwing an exception if not. A different metadataCollection implementation for a different repository that does not support soft delete should not police the status, and instead should go ahead and purge the instance. The methods I think we are talking about here are primarily purgeEntity and purgeRelationship - supported or not by deleteEntity and deleteRelationship - and we are operating on instances that were created locally using addEntity or addRelationship. In all cases, the local metadataCollection is the home for the instance (which is identified by the metadataCollectionId field in the InstanceHeader being set to the metadataCollectionId of the local metadataCollection).

Remote Operations
When an instance is created in a repository an OMRS event is propagated and received at the other repositories in the cohort, (i.e. the receiving repositories are those other than the one at which the instance was created). The event processor at the receiving repository will call saveEntityReferenceCopy or saveRelationshipReferenceCopy. The metadataCollectionId stored in the instance's InstanceHeader will be set to the metadataCollectionId of the repository where the instance was created, hence the receiving metadataCollection processing the event will 'know' that the local repository is not the home repository - now and whenever it retrieves the reference copy from its repository in future.

Suppose now that the instance is deleted at the (remote) home repository. If that repository supports soft delete then (if the paragraph above about local operations is correct) the home repository will insist that a soft delete is performed first. If the home repository does not support soft delete then it would tolerate a direct call to purge (purgeEntity or purgeRelationship). Let's consider both cases:

i) the home repository does not support soft delete. The instance is purged at the home repository and a purge event is propagated. This results in purgeEntityReferenceCopy or purgeRelationshipReferenceCopy being called by the event processor on the other repositories that receive the event. [I am now being guided by what the in-memory metadataCollection does - this is different to my current implementation in Atlas which I think I need to correct.] The metadataCollection's purgeEntityReferenceCopy or purgeRelationshipReferenceCopy should not check whether the reference copy has already been soft deleted - it should just honour the instruction to purge the reference copy. There is no need to check whether the local (event processing) repository is the home repository, because (as Mandy said above) the home repository will (should?) never receive a call to purge a reference copy. 

ii) the home repository does support soft delete. In this case when the instance is soft deleted at the home repository, a delete event is propagated. I think that at each of the other (receiving) metadataCollections/repositories this results in a call to saveEntityReferenceCopy or saveRelationshipReferenceCopy - with the status from the instance in the event (which will be set to DELETED). The reference copy will be transitioned to DELETED state. If the home repository subsequently processes a purge of the instance, a purge event is propagated and the other (receiving) metadataCollections/repositories will be called on purgeEntityReferenceCopy or purgeRelationshipReferenceCopy (the same as in case i) above) and they (as above) will process the purge without checking the status of the instance.

If the above is a correct and complete description, then I need to change the atlas connector slightly as noted in case i) because I am currently checking status during purge of a reference copy.

I also have a question - which is whether on purgeEntityReferenceCopy my connector should always replace the purged entity with a proxy? I notice that the in-memory connector does this, but the atlas connector currently does not.

**Comment by grahamwallis** [comment]
An update to the above - with regard to the final question - I have coded up the addition of an EntityProxy in place of the purged EntityDetail. I wonder whether connectors should only be creating the proxy if the entity to be purged has any relationships. Also wondering whether during purge of a relationship we should be checking for proxies and removing them.

**Comment by cmgrote** [comment]
I have the same understanding as your detailed walkthrough, and it sounds like the "issue" might indeed be the Atlas piece needing to call `purge*ReferenceCopy`.

Good follow-up question: I'm not sure...  I suspect it might come down to whether we expect (rely) on the home repository to notify us of all relationships that should also be changed as a result of the purged entity (so no need for any special handling on receiving side)?

It does seem like the safer option would be to purge all such relationships; is there any risk / value to retaining them as some sort of "hanging" relationship in case the relationship itself and/or the other side have a home repository that *does* support soft deletes?

**Comment by grahamwallis** [comment]
Thanks Chris.
On a home-repository purgeEntity the relationships are purged. So I would expect the home repository to generate purge events for the entity and all its relationships. So at a receiving (non-home) repository, we should see all such events and should purge the corresponding reference copies, but there will be (short) time intervals between the events (and possibly re-ordering or loss of events). 
BTW: I have removed the code I added (yesterday) for creation of a proxy as a result of a purgeEntity, because I think the parallel maintenance of proxies is something that works for the in-memory connector but is not appropriate in atlas. My question about whether, during processing of a purgeRelationship event, I should be purging the proxies becomes a moot point I think.
